WIP: Mothur: update wrappers and add functional tests

shiltemann opened the pull request:
Very much still a work in progress, mothur has 113 tools, so it may take a while, but please feel free to give feedback on anything at any time (or help :) )

TODOs:
- [x] ixnay the python wrapper and add functional tests to all tools
- [x] collection-ify tools (replace --new_datasets with collections and discover_datasets)
- [x] conda dependency
- [ ] update toolshed dependency to match conda version
     - [ ] move blast_legacy package to iuc
- [ ] make some workflows to ship with toolshed (454 and Illumina SOPs)
- [x] add mothur datatypes (see galaxyproject/galaxy/pull/2038)
- [x] add version_command tags
- [ ] add mothur.tre subclass to datatypes in Galaxy
       - [x] check tre format usage in all wrappers
- [x] set minimum required galaxy version to 16.07 (majority of tools will only work with mothur datatypes in Galaxy)
- [ ] update README
- [x] try to reduce size of test-data dir (eliminated 1,5MB)
- [ ] add shed.yml file

Tools:

`*` = with options from metadata (needs mothur datatypes in galaxy)

tools that need collection support added:
- [x] bin.seqs.xml *
- [x] chimera.check.xml
- [x] classify.otu.xml *
- [x] collect.shared.xml *
- [x] collect.single.xml *
- [x] consensus.seqs.xml *
- [x] create.database.xml * (needs output from get.oturep and classify.otu)
- [x] dist.shared.xml
- [x] fastq.info.xml
- [x] filter.seqs.xml
- [x] get.lineage.xml
- [x] get.otulist.xml *
- [x] get.oturep.xml *
- [x] get.sharedseqs.xml
- [x] heatmap.bin.xml *
- [x] heatmap.sim.xml *
- [x] list.otulabels.xml *
- [x] make.biom.xml *
- [x] make.shared_from_biom.xml (removed, can be done in make.shared)
- [x] make.shared.xml *
- [x] metastats.xml *
- [x] normalize.shared.xml *
- [x] otu.association.xml *
- [x] parse.list.xml *
- [x] phylotype.xml
- [x] pca.xml *
- [x] rarefaction.shared.xml *
- [x] rarefaction.single.xml *
- [x] remove.groups.xml
- [x] remove.lineage.xml
- [x] remove.rare.xml *
- [x] sffinfo.xml
- [x] split.abund.xml
- [x] split.groups.xml
- [x] sub.sample.xml *
- [x] summary.shared.xml *
- [x] summary.single.xml *
- [x] tree.shared.xml
- [x] TreeVector.xml (removed, not a mothur tool)
- [x] trim.flows.xml
- [x] trim.seqs.xml
- [x] venn.xml *

others:
- [x] align.check.xml (blast dep)
- [x] align.seqs.xml
- [x] amova.xml
- [x] anosim.xml
- [x] chimera.bellerophon.xml
- [x] chimera.ccode.xml
- [x] chimera.perseus.xml
- [x] chimera.pintail.xml
- [x] chimera.slayer.xml (blast dep)
- [x] chimera.uchime.xml
- [x] chop.seqs.xml
- [x] classify.seqs.xml (blast dep)
- [x] classify.tree.xml
- [x] clearcut.xml
- [x] cluster.classic.xml
- [x] cluster.fragments.xml
- [x] cluster.split.xml
- [x] cluster.xml
- [x] cooccurrence.xml *
- [x] corr.axes.xml *
- [x] count.groups.xml
- [x] count.seqs.xml
- [x] degap.seqs.xml
- [x] deunique.seqs.xml
- [x] deunique.tree.xml
- [x] dist.seqs.xml *
- [x] get.coremicrobiome.xml *
- [x] get.groups.xml
- [x] get.group.xml
- [x] get.otulabels.xml
- [x] get.otus.xml *
- [x] get.rabund.xml *
- [x] get.relabund.xml *
- [x] get.sabund.xml *
- [x] get.seqs.xml
- [x] hcluster.xml
- [x] homova.xml
- [x] indicator.xml *
- [x] libshuff.xml
- [x] list.seqs.xml
- [x] make.contigs.xml
- [x] make.design.xml
- [x] make.fastq.xml
- [x] make.group.xml
- [x] mantel.xml
- [x] merge.files.xml
- [x] merge.groups.xml *
- [x] merge.sfffiles.xml
- [x] nmds.xml
- [x] otu.hierarchy.xml *
- [x] pairwise.seqs.xml (blast dep)
- [x] parsimony.xml
- [x] pcoa.xml
- [x] pcr.seqs.xml
- [x] phylo.diversity.xml
- [x] pre.cluster.xml
- [x] remove.otulabels.xml
- [x] remove.otus.xml *
- [x] remove.seqs.xml
- [x] reverse.seqs.xml
- [x] screen.seqs.xml
- [x] sens.spec.xml
- [x] seq.error.xml
- [x] shhh.flows.xml
- [x] shhh.seqs.xml
- [x] sort.seqs.xml
- [x] summary.qual.xml
- [x] summary.seqs.xml
- [x] summary.tax.xml
- [x] unifrac.unweighted.xml
- [x] unifrac.weighted.xml
- [x] unique.seqs.xml

(new) tools in mothur v36 without wrappers: 

Added:
- [x] merge.taxsummary
- [x] remove.dists
- [x] get.dists, 
- [x] get.label
- [x] get.mimarkspackage
- [x] make.lefse 
- [x] make.lookup 
- [x] make.sra 
- [x] filter.shared, 
- [x] mimarks.attributes 
- [x] primer.design 


Not adding:
- sff.multiple (can be done by supplying collection to sff.info)
- pipeline.pds (is a tutorial pipeline in a single command, can recreate as proper workflow)
- catchall (windows executable)
- sparcc, classify.rf, classify.svm,lefse, mgcluster,kruskal.wallis, get.communitytype (not stable enough; poorly documented, can't find input files that doesn't cause segfault :/ )
- make.table (just alias for count.seqs)
- make.file (not useful inside galaxy)

bgruening commented:
Just awesome!

shiltemann commented:
looks like we can probably do away with the python wrapper altogether; handle the commands entirely in cheetah? (like I did in chimera.check.xml) Anybody have any objections or reasons this may not work in some of the other tools? otherwise I'll just go ahead and do this in all of them, just feels so much cleaner..

bgruening commented:
I'm in favour of this. It eliminates a level of complexity and errors and we can maybe use `argument=` to map to the original shell command .

davebx commented:
I would like to agree with removing the python wrapper. I would like to agree 500%

martenson commented:
looks great Saskia! 
I know nothing about Mothur so please bare with me: is it feasible to reduce the size of the test datasets from 100k lines to something way smaller?

jmchilton commented on tools/mothur/tools/mothur/chimera.check.xml:
```diff
@@ -73,7 +77,18 @@
+            <output name="out_file" file="Mock_S280_L001_R1_001_small.trim.contigs.good.chimeracheck.chimeras" ftype="txt" />
+            <output_collection name="images">
+                <element name="M00967_43_000000000-A3JHG_1_1101_14010_4122.chimeracheck.svg" file="M00967_43_000000000-A3JHG_1_1101_14010_4122.chimeracheck.svg" />
+                <element name="M00967_43_000000000-A3JHG_1_1101_13293_4703.chimeracheck.svg" md5="0d99dc17c183dab4efc75c697060a9a5" />
```
Ideally I'd think you'd remove the ```.chimeracheck.svg`` from the element name I think and it doesn't look like they have a format. So maybe something like:

```
 <discover_datasets pattern="(?P&lt;designation&gt;.*)\.chimeracheck\.svg" directory="images" format="svg" />
```

shiltemann replied:
ah right, I see how that works now, thanks :) btw, is there also a way to test the number of elements in the collection? to ensure I'm getting the expected number of files into the collection?

jmchilton replied:
Not currently possible but I have opened a PR: https://github.com/galaxyproject/galaxy/pull/1942. Hope this helps.

shiltemann commented:
@martenson ah yes, good that you bring it up, been meaning to ask ..there is a 50k line file there that could be removed, but it speeds that particular test up from about a minute to a few seconds, so I don't know what's preferable, smaller test data or faster tests?

the largest by size is HMP_MOCK.v35.align (even though it is only 64 lines), it is like a reference genome, and 1,6 MB is the smallest I could get it without mothur segfaulting or producing empty outputs in a lot of the tools. But it is a mandatory param in about half the tools, so not including it means forgoing test on a lot of the tools :/ If somebody finds a smaller dataset I'll happily replace it, but this is the best I could do for now

davebx commented:
@shiltemann My personal preference when writing tests is that the speed and size of test data are secondary to the accuracy of the tests. So if you need 100k lines to get reliable and correct testing, so be it. If you can improve one over the other, I'd rather have faster tests than smaller data. Others might have very different preferences.

yhoogstrate commented on tools/mothur/tools/mothur/collect.single.xml:
```diff
@@ -5,46 +5,37 @@
-        #end if
+    <command><![CDATA[
+        echo "collect.single(
+            #if $otu.extension == 'mothur.shared':
```
@shiltemann, @jmchilton suggested in another PR to do it a bit different: https://github.com/galaxyproject/tools-iuc/pull/745#discussion_r61588814 - we might need to take a look at this too

shiltemann replied:
ah, great, thanks :)

shiltemann commented:
@IyadKandalaft are you maintaining the blast package (`package_blast_2_2_26`) in the tool shed? the files it downloads from ncbi seem to have moved to a different location, so the installation now fails.

(e.g. ftp://ftp.ncbi.nlm.nih.gov/blast/executables/release/LATEST/blast-2.2.26-src.tar.gz doesn't exist anymore) 

@bgruening should we move this package to iuc as well? or do we want to go conda-only?

bgruening commented:
@shiltemann I'm for moving this package into IUC.
If you ask me we can go for Conda only. But this is not an official IUC guideline. I will discuss is at GCC and hopefully we can do the switch with 16.07.

bgruening commented:
You rock @shiltemann!

shiltemann commented:
@bgruening almost there.. :)

shiltemann commented:
Beauty.

![beauty](https://cloud.githubusercontent.com/assets/2563865/15653628/5e6643ba-268e-11e6-8b1b-8ffd88b67de5.png)

shiltemann commented:
ok, updated all the tools and added 10 new ones. Think this is ready for another round of reviews if anyone is up to it :) 

..since we have to depend on 16.07 for the datatypes anyway, I may hold off on updating the toolshed dependencies until after GCC and see what IUC decides on conda-only dependencies

bgruening commented:
OMG, this is simply beautiful! I will start reviewing this week!
Thanks a bunch!

shiltemann commented:
@bgruening danke shön! (and take your time, I know it's a lot, I'll just leave this to gather review comments and get back to it after GCC :) )

shiltemann commented on tools/mothur/tools/mothur/split.groups.xml:
```diff
@@ -1,44 +1,60 @@
+        <data name="logfile" format="txt" label="${tool.name} on ${on_string}: logfile"/>
+        <collection name="groupfastas" type="list" label="${tool.name} on ${on_string}: fasta per group">
+            <!-- TODO: format_source on discovered collection elements doesn't seem to work yet, change to format_source="fasta" once it does -->
+            <discover_datasets pattern=".*?\.(?P&lt;designation&gt;.*)\.fasta" format="fasta"/>
```
@jmchilton I would like to use format_source here for the collection elements, but that doesn't seem to work? or should I be doing it differently?

jmchilton replied:
Can you put a format_source on the collection tag instead - that works in some contexts? It may be that the discover_datasets overrides that with data or something even if nothing is specified there - in that case I'd open a Galaxy issue and ping me.

shiltemann replied:
tried that too, but no luck (will double check again tomorrow and open an issue if really not) thanks!

martenson commented:
marvelous! :tada: :tada: :tada: 
thanks @shiltemann @bgruening !!!
